### ISSUE TITLE
[HW3-kcov] Where can I find predefined member functions of ASTVisitors?

### ISSUE THREAD

**Issue by Allgot** [swtv-kaist/cs458-spring23#22]
Hello!

Well, I tagged this question as HW3, but it might not be related to it directly.
It is about Clang's ASTVisitor.

In the given sample code, MyASTVisitor is defined as follows:
```
class MyASTVisitor : public RecursiveASTVisitor<MyASTVisitor>
{
public:
...
    bool VisitFunctionDecl(FunctionDecl *f) {
    ...
```

As we can see, the function `VisitFunctionDecl(FunctionDecl *f)` is defined & works well. To check, I defined even more specific function such as `VisitSwitchStmt(SwitchStmt *s)`, and it works well...

However, the template `clang::RecursiveASTVisitor` **does not** define such functions, as [its document](https://clang.llvm.org/doxygen/classclang_1_1RecursiveASTVisitor.html) shows. It only defines somewhat **general** functions including `VisitStmt(Stmt *S)`.

Then how is it possible to such functions actually work?
What is the connecting link between `VisitSwitchStmt(SwitchStmt *s)` and `RecursiveASTVisitor`?

In other words, how the Clang detect such callback functions & invoke them?

**Comment by swtv-kaist** [comment]
Would you share your full example code here to find an answer to your question?  Thanks

**Comment by Allgot** [comment]
Sure!

Since it is very simple, I think there will be no problem to share using the text only.

clang_playground.cpp:
```
(same headers as HW03)
using namespace clang;
using namespace std;

class MyASTVisitor : public RecursiveASTVisitor<MyASTVisitor>
{
public:
   MyASTVisitor(Rewriter &R, const LangOptions &langOptions, SourceManager &SourceMgr)
      : TheRewriter(R), LangOpts(langOptions), m_sourceManager(SourceMgr)
   {}

   bool VisitSwitchStmt(SwitchStmt *s) {
      s->dumpColor();

      return false;
   }

private:
   Rewriter &TheRewriter;
   const LangOptions &LangOpts;
   SourceManager &m_sourceManager;
};

(same MyASTConsumer as HW03)
(same main() as HW03)
```

ex1.c:
```
#include <stdio.h>
int main(void) {
   int a = 0;
   switch (a) {
      case 0:
         break;
   }
}
```

Because `VisitSwitchStmt(SwitchStmt *s)` is the only function has been defined, we can sure that all colored dumps are provided by this function. Therefore, if there is any colored dump, it means that `VisitSwitchStmt(SwitchStmt *s)` is called by Consumer somehow.
And actually, it produces such colored dump.

But the template, `RecursiveASTVisitor` does not contain any information about `VisitSwitchStmt(SwitchStmt *s)`...
Then how does the Consumer invoke `VisitSwitchStmt(SwitchStmt *s)`? And where can I find a list of such `pre-defined` callback functions?

**Comment by swtv-kaist** [comment]
L373 in [```RecursiveASTVisitor.h```](https://clang.llvm.org/doxygen/RecursiveASTVisitor_8h_source.html) defines ```bool VisitSwitchStmt(Switch *S) {return true;}``` via macro.  Clang provides various visit methods for other classes, too.  

```
367 #define STMT(CLASS, PARENT) \
368 bool WalkUpFrom##CLASS(CLASS *S) { \
369 TRY_TO(WalkUpFrom##PARENT(S)); \
370 TRY_TO(Visit##CLASS(S)); \
371 return true; \
372 } \
373 bool Visit##CLASS(CLASS *S) { return true; }
```

However, I recommend to use ```VisitStmt``` instead of ```VisitSwitchStmt``` and so on, since ```VisitStmt``` is an official public method while the others are not.

**Comment by Allgot** [comment]
Oh, so it was not magic but macro...!

Thank you!! 
And yes, it seems better to use `VisitStmt + isa<type>` style as it is explicitly defined & documented ;)
